libraries/AP_Mount/AP_Mount_Topotek.cpp

```diff
@@ -0,0 +1,995 @@
+        return false;
+    }
+
+    distance_m = _measure_dist_m;
+    return true;
+}
+
+// enable/disable rangefinder.  Returns true on success
+bool AP_Mount_Topotek::set_rangefinder_enable(bool enable)
+{
+    if (enable) {
+        _set_gimbal_range_enable[11] = '3';
+    } else {
+        _set_gimbal_range_enable[11] = '0';
+    }
+    return send_packet(_set_gimbal_range_enable, 12);
+}
+
+// get attitude as a quaternion.  returns true on success
+bool AP_Mount_Topotek::get_attitude_quaternion(Quaternion& att_quat)
+{
+    att_quat.from_euler(_current_angle_rad.x, _current_angle_rad.y, _current_angle_rad.z);
+    return true;
+}
+
+// reading incoming packets from gimbal and confirm they are of the correct format
+void AP_Mount_Topotek::read_incoming_packets()
+{
+    // check for bytes on the serial port
+    int16_t nbytes = MIN(_uart->available(), 1024U);
+    if (nbytes <= 0 ) {
+        return;
+    }
+
+    // process bytes received
+    for (int16_t i = 0; i < nbytes; i++) {
+        uint8_t b;
+        if (!_uart->read(b)) {
+            continue;
+        }
+
+        if (_msg_buff_len != 0 || b == '#') {
+            if (b == '#' && _msg_buff_len != 0) {
+                analyse_packet_data();
+            }
+            _msg_buff[_msg_buff_len++] = b;
+        }
+    }
+    analyse_packet_data();
+}
+
+// request gimbal attitude
+void AP_Mount_Topotek::request_gimbal_attitude()
+{
+    send_packet(_get_gimbal_attitude, 12);
+}
+
+// request gimbal memory card information
+void AP_Mount_Topotek::request_gimbal_sdcard_info(void)
+{
+    send_packet(_get_gimbal_sdcard_info, 12);
+}
+
+// request gimbal tracking status
+void AP_Mount_Topotek::request_track_status(void)
+{
+    send_packet(_get_gimbal_track_status, 12);
+}
+
+// request gimbal basic information
+void AP_Mount_Topotek::request_gimbal_basic_info(void)
+{
+    send_packet(_get_gimbal_basic_info, 12);
+}
+
+// set angle target in radians
+void AP_Mount_Topotek::send_angle_target(float roll_rad, float pitch_rad, float yaw_rad, bool yaw_is_ef)
+{
+    // set gimbal's lock state if it has changed
+    if (!set_gimbal_lock(yaw_is_ef)) {
+        return;
+    }
+
+    uint8_t speed = 99;
+    int16_t roll_angle_cd = constrain_int16(degrees(roll_rad) * 100, -18000, 18000);
+    int16_t pitch_angle_cd = constrain_int16(-degrees(pitch_rad) * 100, -9000, 9000);
+    int16_t yaw_angle_cd = constrain_int16(degrees(yaw_rad) * 100, -18000, 18000);
+
+    // control yaw using a gyroscope or magnetic encoding
+    _set_yaw_angle_cmd[8] = yaw_is_ef ? 'I' : 'A';
+
+    _set_yaw_angle_cmd[10] = hex2char((yaw_angle_cd >> 12) & 0x0F);
+    _set_yaw_angle_cmd[11] = hex2char((yaw_angle_cd >> 8) & 0x0F);
+    _set_yaw_angle_cmd[12] = hex2char((yaw_angle_cd >> 4) & 0x0F);
+    _set_yaw_angle_cmd[13] = hex2char((yaw_angle_cd) & 0x0F);
+
+    _set_yaw_angle_cmd[14] = hex2char((speed >> 4) & 0x0F);
+    _set_yaw_angle_cmd[15] = hex2char((speed) & 0x0F);
+
+    send_packet(_set_yaw_angle_cmd, 16);
```

Comment:
maybe send_packet(const char *header, const char *fmt, ...), calls snprintf()
send_packet("#tpUG6wGIY", "%04x%02x", yaw_angle_cd, speed);

Reply:
The header could be a part of the fmt too.

Reply:
The send_packet() method has been replaced by send_variablelen_packet() which uses less flash and better matches our other drivers.